teps/0126-allow-task-sidecars-to-be-specified-in-pipelinerun.md

```diff
@@ -0,0 +1,123 @@
+---
+status: proposed
+title: Allow Task sidecars to be specified in PipelineRun
+creation-date: '2022-11-11'
+last-updated: '2022-11-11'
+authors:
+- '@michaelsauter'
+collaborators: []
+---
+
+# TEP-0126: Allow Task sidecars to be specified in PipelineRun
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+- [Proposal](#proposal)
+  - [Notes and Caveats](#notes-and-caveats)
+- [Design Details](#design-details)
+- [Design Evaluation](#design-evaluation)
+- [Alternatives](#alternatives)
+- [Implementation Plan](#implementation-plan)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Extend `PipelineRun` to allow authors to specify sidecars of tasks.
+
+## Motivation
+
+Make tasks more reusable.
+
+Tasks are meant to be shared (e.g. through Git or the ArtifactHub) and referenced from pipeline runs. This TEP seeks to make tasks more often reusable by removing the limitation that sidecars are "hardcoded" and cannot be changed by the pipeline run author.
+
+### Goals
+
+* Extend `PipelineRun` to allow to specify all aspects of sidecars, matching how [sidecars are specified in tasks](https://tekton.dev/docs/pipelines/tasks/#specifying-sidecars).
+
+### Non-Goals
+
+* Pipeline level sidecars, that live for the duration of the whole pipeline run, see https://github.com/tektoncd/pipeline/issues/2973.
```

Comment:
I know Pipeline-level sidecars are out of scope here, but I'm curious why they wouldn't work for the use case you've identified? The main drawback seems to be that the sidecar lasts for the duration of the PipelineRun, while you only need them for one task. However, this could be combined with [pipelines in pipelines](https://github.com/tektoncd/community/blob/main/teps/0056-pipelines-in-pipelines.md) to create a testing sub-pipeline with a database sidecar.

Reply:
There are two reasons I listed it here:

1. The referenced issue was closed due to lack of interest.
2. I understand pipeline-level sidecars as a completely new feature, while my proposal is only about exposing existing functionality (task-level sidecars) during pipeline authoring time. My hope (without having read the code) is that it would be much easier to add.

Reply:
Thanks! My hope is that we'll prioritize pipeline-level sidecars in the near term, as they're useful for caching docker builds in addition to the integration test use case you've mentioned here. You're right that the feature you're proposing is easier to implement, but we aim to prioritize features that best meet user needs, even if they're a bit harder to create :) I'm just trying to understand here if pipeline-level sidecars are a viable alternative proposal, and it sounds like they are?

Reply:
> My hope is that we'll prioritize pipeline-level sidecars in the near term, as they're useful for caching docker builds in addition to the integration test use case you've mentioned here.

Ah! Did not think of the use case of caching docker builds. That sounds interesting as well.

> You're right that the feature you're proposing is easier to implement, but we aim to prioritize features that best meet user needs, even if they're a bit harder to create :) I'm just trying to understand here if pipeline-level sidecars are a viable alternative proposal, and it sounds like they are?

Makes sense to me. I could live with both pipeline-level or task-level sidecars. Of course, pipeline-level would be a waste of resources for my use case, but then again would be the better option for cases where the sidecar is needed in multiple tasks.

Reply:
FYI, I've opened https://github.com/tektoncd/community/pull/943 proposing pipeline-level sidecars, primarily in the context of docker builds because I think the use case is more clear than for integration tests. LMK if you'd like to collaborate on this!

Reply:
Cool! Looking forward to this. Unfortunately I won't be able to contribute at this point as we had a bit of a shift of priorities in the last month.

Reply:
no problem at all!